checker/mod_checking.ml

```diff
@@ -88,18 +108,32 @@ let mk_mtb mp sign delta =
     mod_delta = delta;
     mod_retroknowledge = ModTypeRK; }
 
-let rec check_module env mp mb =
+let collect_constants_without_body sign mp =
+  let collect_sf s lab = function
+  | SFBconst cb ->
+     let c = Constant.make2 mp lab in
+     if Declareops.constant_has_body cb then s else Cset.add c s
+  | SFBmind _ | SFBmodule _ | SFBmodtype _ -> s in
+  match sign with
+  | MoreFunctor _ -> Cset.empty  (* currently ignored *)
+  | NoFunctor struc ->
+     List.fold_left (fun s (lab,mb) -> collect_sf s lab mb) Cset.empty struc
+
+let rec check_module env opac mp mb =
   Flags.if_verbose Feedback.msg_notice (str "  checking module: " ++ str (ModPath.to_string mp));
   let env = Modops.add_retroknowledge mb.mod_retroknowledge env in
-  let (_:module_signature) =
-    check_signature env mb.mod_type mb.mod_mp mb.mod_delta
+  let sign, opac =
+    check_signature env opac mb.mod_type mb.mod_mp mb.mod_delta Cset.empty
   in
-  let optsign = match mb.mod_expr with
-    |Struct sign -> Some (check_signature env sign mb.mod_mp mb.mod_delta, mb.mod_delta)
-    |Algebraic me -> Some (check_mexpression env me mb.mod_mp mb.mod_delta)
-    |Abstract|FullStruct -> None
+  let optsign, opac = match mb.mod_expr with
+    |Struct sign_struct ->
+      let opacify = collect_constants_without_body sign mb.mod_mp in
+      let sign, opac = check_signature env opac sign_struct mb.mod_mp mb.mod_delta opacify in
+      Some (sign, mb.mod_delta), opac
+    |Algebraic me -> Some (check_mexpression env opac me mb.mod_mp mb.mod_delta), opac
```

Comment:
To deal with the `Module M : T := N` case, it seems that the opac should be extended here with a map of all axioms of the signature to an empty set of dependencies (since the dependencies are all external to the current module).